### PR TITLE
Python UDF

### PR DESCRIPTION
**Pull request by ObjatieGroba**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Python Scalar User Defined Functions

Detailed description / Documentation draft:
Adding Python Scalar User Defined Functions

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

**Comment by ObjatieGroba** [comment]
TODO:
1) Придумать способ определения или создания numpy::dtype по типу UDF
2) Скрестить CREATE FUNCTION с фабрикой

Вощможно стоит перейти на версию сравнимую с Monet DB, а именно не scalar функции использовать, а применимые к одной строке. Только в этом случае можно гарантировать верную запись в нужное место. Тем не менее скорость сильно уменьшится

Также существует пробелма многопоточного исполнение Python. Сейчас используется  mutex. Не ясно как его убрать

File: src/Columns/ColumnDecimal.h [hunk 1]
```diff
@@ -113,7 +113,7 @@ class ColumnDecimal final : public COWHelper<ColumnVectorHelper, ColumnDecimal<T
 
     Field operator[](size_t n) const override { return DecimalField(data[n], scale); }
 
-    StringRef getRawData() const override { return StringRef(reinterpret_cast<const char*>(data.data()), data.size()); }
+    StringRef getRawData() const override { return StringRef(reinterpret_cast<const char*>(data.data()), byteSize()); }
```
Comment (ObjatieGroba) [inline src/Columns/ColumnDecimal.h:116]: Бага это или фича?

Не нашел ни одного употребления size

**Comment by jason-godden** [comment]
This is great! Is there a plan to add simple SQL UDFs though? Essentially like macros? Would be very useful when there are complex formulas that you want to reuse.

**Comment by jason-godden** [comment]
Another question here - would it be easy enough to extend this to allow for anonymous functions? So something like:

```python
SELECT
    id,
    pythonUDF('UInt8', 'parameter1', $$
        import somemodule
        return somemodule.somethingwithparams(parameter1)$$)(someclickhousecolumn)
FROM table
```

Where UInt8 is the return type and parameter1 is the name of the column passed into the function (or however the current implementation will work).

There are so many possibilities with this (thinking with how to integrate this with Vulkn - https://github.com/VulknData/vulkn) - thank you so much for making this happen.

File: src/IO/ReadHelpers.cpp [hunk 2]
```diff
@@ -715,6 +715,29 @@ template void readJSONStringInto<NullSink>(NullSink & s, ReadBuffer & buf);
 template void readJSONStringInto<String>(String & s, ReadBuffer & buf);
 
 
+template <typename Vector>
+bool readBracketStringInto(Vector & s, ReadBuffer & buf)
```
Comment (amosbird) [inline src/IO/ReadHelpers.cpp:719]: It's Brace not Bracket

**Comment by hczhcz** [comment]
@jason-godden 
We have this kind of UDF function in C++ (#6991).
It is not viable in typical scenarios, considering the user code can crash the whole server (no matter if intentionally or unintentionally). We use it to do some quick jobs in our private server.

**Comment by ObjatieGroba** [comment]
@hczhcz
I do research about easiest way of implementation UDF. It is possible to implement isolated C++ UDF  without strong performance degradation (#10750).
I will implement agreed by @alexey-milovidov UDF soon.

**Comment by alexey-milovidov** [comment]
Closed due to zero progress.

**Pull request closed without merging**